File: 4. QP & SQP/optSolver_Woo_Seonho.m
% IOE 511/MATH 562, University of Michigan
% Code written by: Seonho Woo (Claire)

% Function that runs a chosen algorithm on a chosen problem
%           Inputs: problem, method, options (structs)
%           Outputs: final iterate (x) and final function value (f)
function [x, f] = optSolver_Woo_Seonho(problem, method, options)

% set problem, method and options
[problem] = setProblem(problem);
[method] = setMethod(method);
[options] = setOptions(options);

% compute initial function/gradient/Hessian
x = problem.x0;
f = problem.compute_f(x);
g = problem.compute_g(x);
H = problem.compute_H(x);

norm_g = norm(g,inf);
term_constant = max(norm_g, 1);

% set initial iteration counter
k = 0;

while k < options.max_iterations && norm_g >= options.term_tol * term_constant
    
    % take step according to a chosen method
    switch method.name
        
        case 'Newton'
            [x_new,f_new,g_new,H_new,d,alpha,modified] = NewtonStep(x,f,g,H,problem,method,options);

        case 'BFGS'
            if k == 0
            
                H = eye(problem.n);
            end
            [x_new,f_new,g_new,H_new,d,alpha, skipped] = BFGSStep(x,f,g,H,problem,method,options);
            H_old = H;
            H = H_new;
        
        otherwise
            error('Method not implemented yet!')            
    end
    
    % update old and new function values
    x_old = x; f_old = f; g_old = g; norm_g_old = norm_g; H_old = H;
    x = x_new; f = f_new; g = g_new; norm_g = norm(g,inf); H = H_new;

    % increment iteration counter
    k = k + 1;
    
end

end

File: 1. GD & Newton's/optSolver_Woo_Seonho.py
# IOE 511/MATH 562, University of Michigan
# Code written by: Seonho Woo

# Function that runs a chosen algorithm on a chosen problem
#           Inputs: problem, method, options (structs)
#           Outputs: final iterate (x) and final function value (f)
from math import inf
import numpy as np
import functions
import algorithms 

def optSolver_Woo_Seonho(problem,method,options):
    # compute initial function/gradient/Hessian
    x = problem.x0
    f = problem.compute_f(x)
    g = problem.compute_g(x)
    H = problem.compute_H(x)
    norm_g = np.linalg.norm(g, ord=inf)
    g0 = g
    # set initial iteration counter
    k = 0
    f_list = [f]
    #x_list = [x]
    
    while k < options.max_iterations and  norm_g >= options.term_tol*max(np.linalg.norm(g0, ord=inf), 1):
        if method.name == 'GradientDescent':
            x_new,f_new,g_new,d,alpha = algorithms.GDStep(x,f,g,problem,method,options)
            f_list.append(f_new) 

        elif method.name == 'Newton': 
            x_new,f_new,g_new,H_new,d,alpha = algorithms.NewtonStep(x,f,g,H,problem,method,options)
            H_old = H; H = H_new
            f_list.append(f_new) 
            '''
            if isinstance(f_new, list):
                f_list.append(f_new[0])
            else:
                f_list.append(f_new)           
            '''

        else:
            print('Warning: method is not implemented yet')
    
        # update old and new function values        
        x_old = x; f_old = f; g_old = g; norm_g_old = norm_g
        x = x_new; f = f_new; g = g_new; norm_g = np.linalg.norm(g,ord=inf)
        
        # increment iteration counter
        k = k + 1 

    return x,f, f_list
